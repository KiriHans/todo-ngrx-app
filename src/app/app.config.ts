import { ApplicationConfig, isDevMode } from '@angular/core';
import { provideRouter } from '@angular/router';

import { provideAnimations } from '@angular/platform-browser/animations';

import { routes } from './app.routes';
import { provideStore } from '@ngrx/store';
import { provideStoreDevtools } from '@ngrx/store-devtools';
import { appReducers, metaReducers } from './app.reducer';

export const appConfig: ApplicationConfig = {
  providers: [
    provideRouter(routes),
    provideAnimations(),
    provideStore(appReducers, { metaReducers }),
    provideStoreDevtools({ maxAge: 25, logOnly: !isDevMode() }),
  ],
};
