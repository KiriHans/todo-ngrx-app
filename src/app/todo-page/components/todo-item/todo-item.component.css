:host{
	width: 100%;
}

:host ::ng-deep :is(.field-edit-todo) .mat-mdc-form-field-subscript-wrapper {
	display: none;
}

.edit{
  display: none;
}


li{
	list-style: none;
	display: flex;
	justify-content: space-between;
	flex-direction: row;
	align-items: center;
}

.view{
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: center;
	gap: 1rem;

	height: 3.2rem;

	width: 100%;
}

.view label{
	display: flex;
	flex-direction: row;
	align-items: center;

	width: 100%;
	height: 2rem;

	font-size: 1rem;

}

.view mat-checkbox {
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: center;
	transform: scale(1.2);
}


li.editing .view {
	display: none;
}


li.editing .edit {
	font-family: roboto;
	font-size: 1rem;
	display: block;
	width: calc(100% - 80px);
	height: 3rem;
	padding:  0 16px;
	margin: 0 0 0 40px;

	border: 2px solid #ff4081;

}

.edit:focus{
	outline: none;
	box-shadow: 0 0 1px 1px rgba(255, 64, 127, .5);
	
}

.completed{
	color: #949494;
	text-decoration: line-through;
}

.completed .edit {
	color: #949494;
	text-decoration: line-through;

	border: 2px solid hsla(340, 100%, 63%, 0.732);
	
}

.completed .destroy{
	color: #ff4081;
	font-variation-settings:
  'FILL' 0,
  'wght' 600,
  'GRAD' 0,
  'opsz' 48
}