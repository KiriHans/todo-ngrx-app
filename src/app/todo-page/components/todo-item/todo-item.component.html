<li [ngClass]="{ completed: todo.completed, editing: editing }">
  <div class="view">
    <mat-checkbox [formControl]="completed"> </mat-checkbox>

    <label (dblclick)="editTodo()" for="add">
      {{ todo.text }}
    </label>
    <button (click)="deleteTodo()" mat-icon-button class="destroy">
      <mat-icon>close</mat-icon>
    </button>
  </div>

  <input
    [cdkTrapFocusAutoCapture]="editing"
    [cdkTrapFocus]="editing"
    [formControl]="inputTodo"
    [value]="todo.text"
    (blur)="stopEdit()"
    (keyup.enter)="stopEdit()"
    appSelect
    class="edit"
    type="text"
  />
</li>
