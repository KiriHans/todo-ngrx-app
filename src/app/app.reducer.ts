import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import { TodoItem } from './todo-page/models/todo-item.model';
import { todoReducer } from './todo-page/store/reducers/todo.reducer';
import { filterReducer } from './todo-page/store/reducers/filter.reducer';
import { ValidFilters } from './todo-page/types/valid-filter.type';
import { storageMetaReducer } from './todo-page/store/reducers/todo.metareducer';

export interface AppState {
  todos: TodoItem[];
  filter: ValidFilters;
}

export const appReducers: ActionReducerMap<AppState> = {
  todos: todoReducer,
  filter: filterReducer,
};

export const metaReducers: MetaReducer[] = [storageMetaReducer];
