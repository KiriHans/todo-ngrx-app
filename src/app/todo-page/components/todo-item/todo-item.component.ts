import { Component, DestroyRef, Input, OnInit, inject } from '@angular/core';
import { CommonModule } from '@angular/common';
import { A11yModule } from '@angular/cdk/a11y';
import { FormControl, ReactiveFormsModule, Validators } from '@angular/forms';
import { SelectInputDirective } from '../../directives/select-input.directive';

/* RXJS Modules */
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';

/* NGRX ModulesÂ¨*/
import { Store } from '@ngrx/store';
import * as TodoActions from '../../store/actions/todo.actions';

/* Angular Material Modules */
import { MatInputModule } from '@angular/material/input';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';

import { TodoItem } from '../../models/todo-item.model';

@Component({
  selector: 'app-todo-item',
  standalone: true,
  imports: [
    CommonModule,
    A11yModule,
    ReactiveFormsModule,
    SelectInputDirective,
    MatInputModule,
    MatCheckboxModule,
    MatButtonModule,
    MatIconModule,
  ],
  templateUrl: './todo-item.component.html',
  styleUrls: ['./todo-item.component.css'],
})
export class TodoItemComponent implements OnInit {
  @Input() todo!: TodoItem;

  private destroyRef = inject(DestroyRef);

  editing = false;
  show = false;

  completed!: FormControl;
  inputTodo!: FormControl;

  constructor(private store: Store) {}

  ngOnInit(): void {
    this.completed = new FormControl(this.todo.completed);
    this.inputTodo = new FormControl(this.todo.text, Validators.required);

    this.completed.valueChanges.pipe(takeUntilDestroyed(this.destroyRef)).subscribe({
      next: () => {
        this.store.dispatch(TodoActions.toggleTodo({ id: this.todo.id }));
      },
    });
  }

  editTodo(): void {
    this.editing = true;
    this.show = true;

    this.inputTodo.setValue(this.todo.text);
  }

  stopEdit(): void {
    this.editing = false;
    this.show = false;

    if (this.inputTodo.invalid || this.inputTodo.value === this.todo.text) return;

    this.store.dispatch(TodoActions.editTodo({ id: this.todo.id, newText: this.inputTodo.value }));
  }

  deleteTodo(): void {
    this.store.dispatch(TodoActions.deleteTodo({ id: this.todo.id }));
  }
}
