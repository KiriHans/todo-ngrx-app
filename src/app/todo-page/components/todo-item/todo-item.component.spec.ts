import { ComponentFixture, TestBed } from '@angular/core/testing';

import { HarnessLoader } from '@angular/cdk/testing';
import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
import { MatCheckboxHarness } from '@angular/material/checkbox/testing';

import { TodoItemComponent } from './todo-item.component';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { AppState } from 'src/app/app.reducer';
import { TodoItem } from '../../models/todo-item.model';
import { A11yModule } from '@angular/cdk/a11y';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { SelectInputDirective } from '../../directives/select-input.directive';
import { Action } from '@ngrx/store';
import { MatButtonHarness } from '@angular/material/button/testing';

describe('TodoItemComponent', () => {
  let component: TodoItemComponent;
  let fixture: ComponentFixture<TodoItemComponent>;
  let el: DebugElement;
  let store: MockStore<AppState>;

  let compiled: HTMLElement;

  let harnessLoader: HarnessLoader;

  let dispatchSpy: jest.SpyInstance<void, [action: Action], unknown>;

  const todoItemTest = new TodoItem('test');

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        TodoItemComponent,
        NoopAnimationsModule,
        A11yModule,
        ReactiveFormsModule,
        SelectInputDirective,
      ],
      providers: [provideMockStore()],
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TodoItemComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;
    compiled = fixture.nativeElement;

    harnessLoader = TestbedHarnessEnvironment.loader(fixture);

    component.todo = todoItemTest;
  });

  beforeEach(() => {
    store = TestBed.inject(MockStore);
    dispatchSpy = jest.spyOn(store, 'dispatch');

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should render a li element', () => {
    const li = el.query(By.css('li'));

    expect(li).toBeTruthy();
  });

  it('should render an element with "view" class containing a checkbox and a label', () => {
    const viewElement = el.query(By.css('.view'));

    expect(viewElement).toBeTruthy();

    const checkbox = viewElement.query(By.css('mat-checkbox'));
    const label = viewElement.query(By.css('label[for="edit"]'));

    expect(checkbox).toBeTruthy();

    expect(label).toBeTruthy();
    expect(label.nativeElement.textContent.trim()).toBe(todoItemTest.text);
  });

  it('should render an input element with "edit" class', () => {
    const input = el.query(By.css('input.edit'));

    expect(input).toBeTruthy();
  });

  it('should initialize correctly', () => {
    expect(component.completed).toBeInstanceOf(FormControl);
    expect(component.inputTodo).toBeInstanceOf(FormControl);
  });

  it('should dispatch an action when checkbox is pressed', async () => {
    const checkbox = await harnessLoader.getHarness(MatCheckboxHarness);

    expect(dispatchSpy).not.toHaveBeenCalled();

    await checkbox.check();

    expect(dispatchSpy).toHaveBeenCalled();
  });

  it('should change class of li element to "editing" when label is double-clicked', () => {
    const editTodoSpy = jest.spyOn(component, 'editTodo');
    let editing = el.query(By.css('.editing'));
    const label = el.query(By.css('label[for="edit"]'));

    expect(editing).toBeFalsy();

    label.triggerEventHandler('dblclick');

    fixture.detectChanges();

    editing = el.query(By.css('.editing'));

    expect(editTodoSpy).toHaveBeenCalled();
    expect(editing).toBeTruthy();
  });

  it('should reset Form Control inputTodo if input value is empty and to not dispatch any action', () => {
    const stopEditSpy = jest.spyOn(component, 'stopEdit');

    const label = el.query(By.css('label[for="edit"]'));
    const input = el.query(By.css('input.edit'));

    label.triggerEventHandler('dblclick');

    component.inputTodo.setValue('');

    fixture.detectChanges();

    input.triggerEventHandler('blur');

    fixture.detectChanges();
    expect(stopEditSpy).toHaveBeenCalled();
    expect(dispatchSpy).not.toHaveBeenCalled();

    label.triggerEventHandler('dblclick');

    expect(component.inputTodo.value).toBe(todoItemTest.text);
  });

  it("should not dispatch any action if input value hasn't changed", () => {
    const stopEditSpy = jest.spyOn(component, 'stopEdit');

    const label = el.query(By.css('label[for="edit"]'));
    const input = el.query(By.css('input.edit'));

    label.triggerEventHandler('dblclick');

    fixture.detectChanges();

    input.triggerEventHandler('blur');

    expect(stopEditSpy).toHaveBeenCalled();
    expect(dispatchSpy).not.toHaveBeenCalled();
  });

  it('should dispatch an action when input value changes', () => {
    const stopEditSpy = jest.spyOn(component, 'stopEdit');

    const label = el.query(By.css('label[for="edit"]'));
    const input = el.query(By.css('input.edit'));

    label.triggerEventHandler('dblclick');
    component.inputTodo.setValue('new Value');

    fixture.detectChanges();

    input.triggerEventHandler('blur');

    expect(stopEditSpy).toHaveBeenCalled();
    expect(dispatchSpy).toHaveBeenCalled();
  });

  it('should trigger stopEdit function when Enter Key is pressed with a correct value', () => {
    const stopEditSpy = jest.spyOn(component, 'stopEdit');

    const label = el.query(By.css('label[for="edit"]'));
    const input = el.query(By.css('input.edit'));

    label.triggerEventHandler('dblclick');
    component.inputTodo.setValue('new Value');

    fixture.detectChanges();

    input.triggerEventHandler('keyup.enter');

    expect(stopEditSpy).toHaveBeenCalled();
    expect(dispatchSpy).toHaveBeenCalled();
  });

  it('should change Form Control value if checkbox is pressed', async () => {
    expect(component.completed.value).toBe(false);

    const checkbox = await harnessLoader.getHarness(MatCheckboxHarness);

    await checkbox.check();

    expect(component.completed.value).toBe(true);
  });

  it('should dispatch an action if "destroy" button is pressed', async () => {
    const deleteTodoSpy = jest.spyOn(component, 'deleteTodo');
    const button = await harnessLoader.getHarness(MatButtonHarness.with({ selector: '.destroy' }));

    await button.click();

    expect(deleteTodoSpy).toHaveBeenCalled();
    expect(dispatchSpy).toHaveBeenCalled();
  });
});
