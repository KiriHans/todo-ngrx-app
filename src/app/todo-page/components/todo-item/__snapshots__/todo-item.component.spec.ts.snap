// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItemComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <li>
    <div
      class="view"
    >
      <mat-checkbox
        class="mat-mdc-checkbox mat-accent _mat-animation-noopable ng-untouched ng-pristine ng-valid"
      >
        <div
          class="mdc-form-field"
        >
          <div
            class="mdc-checkbox"
          >
            <div
              class="mat-mdc-checkbox-touch-target"
            />
            <input
              class="mdc-checkbox__native-control"
              id="mat-mdc-checkbox-2-input"
              tabindex="0"
              type="checkbox"
            />
            <div
              class="mdc-checkbox__ripple"
            />
            <div
              class="mdc-checkbox__background"
            >
              <svg
                aria-hidden="true"
                class="mdc-checkbox__checkmark"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  class="mdc-checkbox__checkmark-path"
                  d="M1.73,12.91 8.1,19.28 22.79,4.59"
                  fill="none"
                />
              </svg>
              <div
                class="mdc-checkbox__mixedmark"
              />
            </div>
            <div
              class="mat-ripple mat-mdc-checkbox-ripple mat-mdc-focus-indicator"
              mat-ripple=""
            />
          </div>
          <label
            class="mdc-label"
            for="mat-mdc-checkbox-2-input"
          />
        </div>
      </mat-checkbox>
      <label
        for="edit"
      >
         test 
      </label>
      <button
        class="destroy mdc-icon-button mat-mdc-icon-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
        mat-button-ripple-uninitialized=""
        mat-icon-button=""
      >
        <span
          class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
        />
        <mat-icon
          aria-hidden="true"
          class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
          data-mat-icon-type="font"
          role="img"
        >
          close
        </mat-icon>
        <span
          class="mat-mdc-focus-indicator"
        />
        <span
          class="mat-mdc-button-touch-target"
        />
      </button>
    </div>
    <div
      aria-hidden="true"
      class="cdk-visually-hidden cdk-focus-trap-anchor"
    />
    <input
      appselect=""
      class="edit ng-untouched ng-pristine ng-valid"
      type="text"
    />
    <div
      aria-hidden="true"
      class="cdk-visually-hidden cdk-focus-trap-anchor"
    />
  </li>
</div>
`;
